src/main/java/com/google/sps/data/PersistentMatchRepository.java

```diff
@@ -0,0 +1,183 @@
+  
+  /**
+  * Constructor initializes private variables.
+  */
+  public PersistentMatchRepository() {
+      datastore = DatastoreServiceFactory.getDatastoreService();
+      gson = new Gson();
+      userRepository = new PersistentUserRepository();
+  }
+
+ /**
+  * Private method to add an entity to the datastore, or to override an existing 
+  * entity when new matches are added.
+  */
+  private void addNewEntity(String userId, String matchString) {
+    Entity userEntity = new Entity("User", userId);
+
+    userEntity.setProperty("userId", userId);
+    userEntity.setProperty("matchIds", matchString);
+    datastore.put(userEntity);
+  }
+
+ /**
+  * Private method to modify the json string and add a match to an existing user.
+  */
+  private void addMatchToExistingUser(String userId, String matchId) {
+    Collection<String> currentMatches = getMatchIdsForUser(userId);
+    currentMatches.add(matchId);
+    String matchString = gson.toJson(currentMatches);
+    addNewEntity(userId, matchString);
+  }
+
+  /**
+  * Private method that returns a Map of all the users and saved matches fitting 
+  * PreparedQuery results.
+  */
+  private Map<String, Collection<String>> fetchMapFromQuery(PreparedQuery results) {
+    Map<String, Collection<String>> idMap = new HashMap<String, Collection<String>>();
+
+    for (Entity entity : results.asIterable()) {
+        String userId = (String) entity.getProperty("userId");
+        String matchIds = (String) entity.getProperty("matchIds");
+        Collection<String> matchCollection = gson.fromJson(matchIds, HashSet.class);
+
+        idMap.put(userId, matchCollection);
+    }
+
+    return idMap;
+  }
+
+  /**
+  * Private method to fetch the full ID Map from the database.
+  */
+  private Map<String, Collection<String>> fetchFullMap() {
+    Query query = new Query("User");
+    PreparedQuery results = datastore.prepare(query);
+    Map<String, Collection<String>> idMap = fetchMapFromQuery(results);
+    return idMap;
+  }
+
+  /**
+  * Private method that sets a query filter based on User ID for the results in the datastore.
+  */
+  private PreparedQuery getQueryFilterForId(String userId) {
+    Filter userNameFilter = new FilterPredicate("userId", FilterOperator.EQUAL, userId);
+    Query query = new Query("User").setFilter(userNameFilter);
+    PreparedQuery results = datastore.prepare(query);
+    return results;
+  }
+
+  /**
+  * Private method that fetches a single user's match map given their User ID.
+  */
+  private Map<String, Collection<String>> fetchMapFromId(String userId) {
+    PreparedQuery results = getQueryFilterForId(userId);
+    Map<String, Collection<String>> userMap = fetchMapFromQuery(results);
+
+    return userMap;
+  }
+
+
+  public void addMatch(User user, User match) {
+    String userId = user.getId();
+    String matchId = match.getId();
+    Map<String, Collection<String>> userMap = fetchMapFromId(userId);
+
+    if(userMap.size() == 0) { //user is not already saved
```

Comment:
nit: Prefer using isEmpty() for improved readability and performance.

Reply:
Done.